src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriter.java

```diff
@@ -0,0 +1,295 @@
+import org.opensearch.knn.indices.Model;
+import org.opensearch.knn.indices.ModelCache;
+import org.opensearch.knn.plugin.stats.KNNGraphValue;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.lucene.codecs.CodecUtil.FOOTER_MAGIC;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.opensearch.knn.common.FieldInfoExtractor.extractKNNEngine;
+import static org.opensearch.knn.common.FieldInfoExtractor.extractVectorDataType;
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+import static org.opensearch.knn.common.KNNConstants.PARAMETERS;
+import static org.opensearch.knn.index.codec.util.KNNCodecUtil.buildEngineFileName;
+import static org.opensearch.knn.index.engine.faiss.Faiss.FAISS_BINARY_INDEX_DESCRIPTION_PREFIX;
+
+/**
+ * Writes KNN Index for a field in a segment. This is intended to be used for native engines
+ */
+@AllArgsConstructor
+@Log4j2
+public class NativeIndexWriter {
+    private static final Long CRC32_CHECKSUM_SANITY = 0xFFFFFFFF00000000L;
+
+    private final SegmentWriteState state;
+    private final FieldInfo fieldInfo;
+    private final NativeIndexBuildStrategy indexBuilder;
+
+    /**
+     * Gets the correct writer type from fieldInfo
+     *
+     * @param fieldInfo
+     * @return correct NativeIndexWriter to make index specified in fieldInfo
+     */
+    public static NativeIndexWriter getWriter(final FieldInfo fieldInfo, SegmentWriteState state) {
+        final KNNEngine knnEngine = extractKNNEngine(fieldInfo);
+        boolean isTemplate = fieldInfo.attributes().containsKey(MODEL_ID);
+        boolean iterative = !isTemplate && KNNEngine.FAISS == knnEngine;
+        if (iterative) {
+            return new NativeIndexWriter(state, fieldInfo, MemOptimizedNativeIndexBuildStrategy.getInstance());
+        }
+        return new NativeIndexWriter(state, fieldInfo, DefaultIndexBuildStrategy.getInstance());
+    }
+
+    /**
+     * flushes the index
+     *
+     * @param knnVectorValues
+     * @throws IOException
+     */
+    public void flushIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
+        knnVectorValues.init();
+        buildAndWriteIndex(knnVectorValues);
+        recordRefreshStats();
+    }
+
+    /**
+     * Merges kNN index
+     * @param knnVectorValues
+     * @throws IOException
+     */
+    public void mergeIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
+        knnVectorValues.init();
+        if (knnVectorValues.docId() == NO_MORE_DOCS) {
+            // This is in place so we do not add metrics
+            return;
+        }
+
+        long arraySize = knnVectorValues.bytesPerVector();
+        startMergeStats(knnVectorValues.dimension(), arraySize);
```

Comment:
Here you need knnVectorValues.totalLiveDocs and not dimension. because method signature is:
```
 private void startMergeStats(int numDocs, long arraySize) {}
```

Reply:
Its confusing because of [this](https://github.com/opensearch-project/k-NN/pull/1950/files#diff-75a39a358693703da64ecd93181897b1e20ad25f951f0f3813148c21257a999cR244) where it says bytes

```
private void startMergeStats(int numDocs, long arraySize) {
        KNNGraphValue.MERGE_CURRENT_OPERATIONS.increment();
        KNNGraphValue.MERGE_CURRENT_DOCS.incrementBy(numDocs);
        KNNGraphValue.MERGE_CURRENT_SIZE_IN_BYTES.incrementBy(arraySize);
        KNNGraphValue.MERGE_TOTAL_OPERATIONS.increment();
        KNNGraphValue.MERGE_TOTAL_DOCS.incrementBy(numDocs);
        KNNGraphValue.MERGE_TOTAL_SIZE_IN_BYTES.incrementBy(arraySize);
    }
```
need clarification here